src/FormParserFactory.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace React\Http;
+
+use React\Http\Parser\FormUrlencoded;
+use React\Http\Parser\Multipart;
+use React\Http\Parser\NoBody;
+use React\Http\Parser\ParserInterface;
+
+class FormParserFactory
+{
+    /**
+     * @param Request $request
+     * @return ParserInterface
+     */
+    public static function create(Request $request)
+    {
+        $headers = $request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        if (!array_key_exists('content-type', $headers)) {
+            return new NoBody($request);
+        }
+
+        $contentType = strtolower($headers['content-type']);
+
+        if (strpos($contentType, 'multipart/') === 0) {
+            return new Multipart($request);
+        }
+
+        if (strpos($contentType, 'application/x-www-form-urlencoded') === 0) {
+            return new FormUrlencoded($request);
+        }
+
+        return new NoBody($request);
```

Comment:
Here is a quick hack up how to deal with JSON in the current situation:

```php
<?php

use React\Http\DeferredStream;
use React\Http\FormParserFactory;
use React\Http\Parser\NoBody;
use React\Promise\Deferred;

require 'vendor/autoload.php';

$loop = \React\EventLoop\Factory::create();
$socket = new React\Socket\Server($loop);

$http = new React\Http\Server($socket);
$http->on('request', function ($request, $response) {
    $parser = FormParserFactory::create($request);
    if ($parser instanceof NoBody) {
        $deferred = new Deferred();
        $promise = $deferred->promise();
        $buffer = '';
        $request->on('data', function ($data) use (&$buffer, $deferred, $request) {
            $buffer .= $data;
            $json = json_decode($buffer);
            if (json_last_error() == JSON_ERROR_NONE) {
                $request->removeAllListeners('data');
                $deferred->resolve($json);
            }
        });
    } else {
        $promise = DeferredStream::create($parser);
    }
    $promise->then(function ($result) use ($response) {
        var_export($result);
        $response->writeHead(200, array('Content-Type' => 'text/plain'));
        $response->end("Hello World!\n");
    });
});

$socket->listen(1337);
$loop->run();
```

`curl -H "Content-Type: application/json" -X POST -d '{"username":"xyz","password":"xyz"}' http://localhost:1337`

The main concern I have is that there is currently no way to detect the en of the body if content length is missing. Also in in the light of your other comment, dropping `NoBody` and throwing an exception instead could be an option. What is your take on doing that?